### DISCUSSION TITLE
Access the root element's context in gatsby-node.js

### DISCUSSION THREAD

**Discussion by yidongw** [gatsbyjs/gatsby#31197] [Help]
I'm using mobx-state-tree for storing data, and it contains some product data that doesn't change very often, so I want to get those data ready in the build time.

I implemented `wrapRootElement` in gatsby-ssr.js and gatsby-browser.js. `wrapRootElement` is the same function that creates a store and wrapping the element with context provider
```
import { MainContext, MainModel } from "../stores/MainStore"

export default ({ element }) => {
  const store = MainModel.createStore({})
  return <MainContext.Provider value={store}>{element}</MainContext.Provider>
}
```
I also need to `createPage` in gatsby-node.js according to [using-gatsby-without-graphql](https://www.gatsbyjs.com/docs/how-to/querying-data/using-gatsby-without-graphql/), but all the data fetching is currently done in the store, so in order to fetch the data, I need to somehow access the store that was created in the gatsby-ssr.js or gatsby-browser.js.

How am I supposed to do that? Or I should use a different approach?

**Comment by Js-Brecht** [comment]
@yidongw let me see if I am understanding correctly:

You have data fetching logic defined in your store, and you want to prefetch that data during build time so that it is available during runtime (without the FE needing to refetch it); and you want to do this without duplicating the data fetching logic in order to use it from node.

First of all, you cannot access a store instance that you've created in `gatsby-ssr`/`gatsby-browser` from within `gatsby-node`.  They're running in different contexts / environments.

You can, however, use the same store logic in `gatsby-node` that you would use in runtime if you're okay with using something like the [esm](https://www.npmjs.com/package/esm) library. You just need make sure your store code doesn't use any special syntax (like jsx) or anything that has to be used in the browser; also need to make sure your store modules don't _import_ any other modules that do this either.  Another way is to use something like [babel-register](https://www.npmjs.com/package/babel-register) or [ts-node](https://www.npmjs.com/package/ts-node) (if you're using typescript) for transpiling modules during import, so that the code can be converted into something usable by node prior to being consumed.  My plugin [gatsby-plugin-ts-config](https://www.npmjs.com/package/gatsby-plugin-ts-config) does this (and your code doesn't need to be written in Typescript when using it, either).

By doing this, you can create a store instance in node and run the data fetching api.  You will then be creating a new store instance inside of `gatsby-ssr`/`gatsby-browser` and using the data you fetched in `gatsby-node` to hydrate the new store instance.  There's a few different ways to get make this data available.

You could:

* Use `createPage` to add the collected data to page context
  * Downside to this is if this data is used globally, you will wind up attaching it to the context of every single page so that it is available no matter which page the user lands on.  This means that it will be included in every single `page-data` chunk, and will cause it to be redownloaded on every page change.

* Push your collected data into the GraphQL layer using `sourceNodes`.
  * Downside is that you now need to explicitly define the shape of your data in a GraphQL query in your react tree, much like the example in #23239.
  * _I'm assuming you don't want to do it this way because you've mentioned doing it without GraphQL._
 
* Store the data in a `json` file somewhere, and define that path in a webpack alias so that you can use something like this inside of `gatsby-ssr` & `gatsby-browser`:
  ```js
  const data = require("__some_webpack_alias__");

  // With a webpack alias, you can even do a ES6 import:
  import data from "__some_webpack_alias__";
  ```
  * This will bundle the data into your javascript, which will make your initial download larger... but you will need to download the data  during hydration anyway, even with the GraphQL method :man_shrugging: 

* You could even store the data in a json format inside of the html of each page (inside of a `<script type="application/json>` tag)
  * This increases the size of your client's initial html download, but it is only going to be done once.
  * This may also affect other metrics, like TTFP

Probably one of the best ways is to push the data into the GraphQL layer because using a static query means your data will be available during SSR and it will also be split into its own data chunk for consumption in the browser, reducing your initial bundle size.  This can lead to more maintenance though due to the nature of GraphQL, so I understand if you want to avoid it.

IIRC, there's a way to do the webpack alias method that will let you load a json chunk synchronously during SSR but also load it in its own chunk (asynchronously) during runtime, much like the static query process does.  It's been a while since I've tried to do that though, so I'd have to try to put together a POC to even remember how.

If any of these methods sounds appealing or if I've misunderstood what you're trying to accomplish, let me know.  If you want to move forward with one of these, maybe you can share your project and we can make it work.

**Reply by yidongw** [reply]
@Js-Brecht Thank you so much for answering my question. I really appreciate it.

Currently, I have a `mainStore` and it has several sub-stores, one of which is `productsStore`. Other sub-stores are relying on `productsStore`. `productsStore` doesn't just have the data, it also has action to fetch the data and then store it in the `productsStore` and functions to get modified data from the original products data.

I'm storing the `mainStore` in the context, so all my components can access it.

What I want to do during the build time is 1. fetch the products data, 2. store it in the `productsStore` and 3. set context value as `mainStore`, then 4. my components and other sub-store can consume `productsStore`.

Your solution is really helpful.

For fetching the data, what API function should I use in `gatsby-node.js`?
In there, I probably would 1. create the store, 2. doing async action for fetching data and save it in the store, and 3. save the snapshot of this store as `json` file somewhere. I also need to make it executable somehow like you mentioned.

Does gatsby execute `gatsby-node.js` before `gatsby-ssr.js` and `gatsby-browser.js`?
My data is relatively small, so it wouldn't increase the size that much. In `gatsby-ssr.js` and `gatsby-browser.js` I can re-create the store with the snapshot I got from `gatsby-node.js` as an argument and use it as context.

How do you like this approach?

**Reply by yidongw** [reply]
@Js-Brecht Also found this [PR](https://github.com/gatsbyjs/gatsby/pull/2423), not sure if it is possible to async fetch data in `gatsby-ssr.js` and `gatsby-browser.js`. Or is it better to do it in `gatsby-node.js`?

Also was considering using this [package](https://github.com/d4rky-pl/arboris), but how sure how to integrate it into gatsby

**Reply by Js-Brecht** [reply]
> For fetching the data, what API function should I use in `gatsby-node.js`?

If all you want to do is save the data as a json file, then you could do it inside of `onPreInit`, `sourceNodes`, `createPages`, `onPreBuild`.  These all run before the build.  Traditionally, data fetching/processing is done within `sourceNodes` because it is commonly pushed into GraphQL.  Additional work was recently done to allow multiple `sourceNodes` apis to run in parallel, so you may get performance benefits by doing it there, too.

Another sensible option would be to do it inside of `createPages` if you wanted to use page context.  For example, if you have some global data you want to store for the whole site, but you also want to create individual pages with their own unique context... it would make sense to do it all in one place.  The GraphQL layer is available to use in `createPages` too, so you could even fetch data from other source plugins and use it for your own api call

> How do you like this approach?

That makes sense.  You can use [onCreateWebpackConfig](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#onCreateWebpackConfig) to modify the webpack config in order to add an alias to your json snapshot.  Import that using the ES6 import I mentioned and your data will be consumable during SSR and in the browser.

> Does gatsby execute `gatsby-node.js` before `gatsby-ssr.js` and `gatsby-browser.js`?

Technically speaking, `gatsby-node` runs before, during, and after `gatsby-ssr`/`gatsby-browser`.  API endpoints defined inside of `gatsby-node` are run at different points in the lifecycle.  `gatsby-ssr`/`gatsby-browser` are bundled by Webpack, `gatsby-ssr` is used during the SSR process (`React.renderToString`), and `gatsby-browser` is included in the client-side bundle.

> Also found this [PR](https://github.com/gatsbyjs/gatsby/pull/2423), not sure if it is possible to async fetch data in `gatsby-ssr.js` and `gatsby-browser.js`. Or is it better to do it in `gatsby-node.js`?

AFAIK, it is not possible yet.  `gatsby-ssr` api endpoints do not support async processing yet, so it is better to do it in `gatsby-node` for now.

Just a side note: I am not a member of the core team, so I don't know what the plans are regarding async ssr; I'm also not nearly as active here as I once was, so I'm not as up to date on current development.  Since I'm a bit out of date, it may be worthwhile to follow up elsewhere, maybe open an issue/discussion regarding this.  Or you can look in the code under `packages/gatsby/cache-dir/api-runner-ssr.js`

> Also was considering using this [package](https://github.com/d4rky-pl/arboris), but how sure how to integrate it into gatsby

You would use the [replaceRenderer](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-ssr/#replaceRenderer) for that, but it wouldn't work because async processing isn't supported.

**Reply by yidongw** [reply]
Thank you so much for answering all that! Now I know how to proceed.

One question, if I only have one page, what is the difference between creating a root element with context using `createPages` in `gatsby-node` VS. using `wrapRootElement` in `gatsby-ssr.js` and `gatsby-browser.js`?

I think I need to have a `json` file either way because you can't pass data between two APIs. I'm just curious about the difference between them. Maybe there are some benefits of choosing one vs. the other.

**Reply by Js-Brecht** [reply]
`wrapRootElement` wraps the root element on _all_ pages.  The `createPages` api just adds the data needed to tell Gatsby what pages to create.  The reason why putting the json import in `gatsby-ssr` and `gatsby-browser` will bundle it for your entire app is because those files are shared across every page... so it becomes a part of a common bundle.

The data you add to context in `createPages` gets added to a `page-data` json chunk, which Gatsby will download when somebody navigates to a page (along with prefetching to make navigating fast).  These are unique per page, and contain the context you provided to `createPage()`.  When the page is rendered, the data from the json chunk is passed to it as a prop.

So really, if you only have a single page, it will likely be easier to use context from `createPage()`.  You can use that same page data from the page props inside of `wrapRootElement` to hydrate your store.  You will just need to remember how this will affect your site if you add more pages later

**Reply by yidongw** [reply]
It is working now! I would be happy to write up an example if you think it's valuable.

For the `json` file, do I need to remove it after it has been imported? I think it should be okay to leave it there as later the `gatsby-node` would overwrite it.

I don't want to put the `json` file into git but typescript would complain it is not there. Haven't thought of a good solution for that yet.

I was trying to make the gatsby config files into typescript but I wasn't able to find the types library for gatsby API, so I give up. [This](https://github.com/alvis/gatsby-webapp-in-typescript) is the closest thing I can find

**Reply by Js-Brecht** [reply]
@yidongw sorry for the late response; I thought I had responded to your message already.

> For the `json` file, do I need to remove it after it has been imported? I think it should be okay to leave it there as later the `gatsby-node` would overwrite it.

These kind of files are commonly kept in the cache, which you can use via the `cache` object in the first parameter of the `gatsby-node` apis:

```js
const path = require("path");
const fs = require("fs-extra"); // <-- handy fs like library that supports promises

const getDataFp = (cache) => path.join(cache.directory, "data.json");

exports.sourceNodes = async ({ cache, reporter }) => {
  const data = ...
  const dataFp = getDataFp();
  await fs.ensureDir(path.dirname(dataFp));
  await fs.writeFile(getDataFp(cache), data);
}

exports.onCreateWebpackConfig = async ({ actions, cache }) => {
  const { setWebpackConfig } = actions;
  setWebpackConfig({
    resolve: {
      alias: {
        "__my_data_file__": getDataFp(cache)
      }
    }
  });
}
```

This also means that a `gatsby clean` will remove them, and they also will not be committed to source control.

> I was trying to make the gatsby config files into typescript but I wasn't able to find the types library for gatsby API, so I give up. [This](https://github.com/alvis/gatsby-webapp-in-typescript) is the closest thing I can find

I have seen that project, and it is a valid way to do it.  I have this plugin to make it easy: [gatsby-plugin-ts-config](https://www.npmjs.com/package/gatsby-plugin-ts-config).  It does a little bit extra, too.  v2 coming soon: https://github.com/Js-Brecht/gatsby-plugin-ts-config/issues/31

Gatsby's types are exported from the index: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/index.d.ts
So you can do something like this:
```ts
import { GatsbyNode } from "gatsby";

export const sourceNodes: GatsbyNode["sourceNodes"] = ({ actions }) => {
  ...
}
```
Note: _There are some issues in the types because of some invalid overloads.  Should be fixed in an [upcoming release](https://github.com/gatsbyjs/gatsby/pull/30819#issuecomment-833388813)_

**Reply by Js-Brecht** [reply]
> I don't want to put the `json` file into git but typescript would complain it is not there. Haven't thought of a good solution for that yet.

In my previous example, I used a webpack alias like `__my_data_file__`.  Obviously you can call it whatever you want, but that's where I would start:

```ts
// some-global.d.ts file
declare module "__my_data_file__" {
  const data: Record<string, unknown>; // Or whatever the type is of your data
  export default data;
}
```
```ts
// gatsby-browser/ssr.ts
import storeSnapshot from "__my_data_file__";
```
There are libraries that will generate interfaces from json, if you wanted to have types based on the data file.  For example: https://github.com/quicktype/quicktype#calling-quicktype-from-javascript or https://github.com/shakyShane/json-ts#usage-api

> I would be happy to write up an example if you think it's valuable

It could be useful for others coming along in the future :man_shrugging: .  You could open a PR to save an example in `/examples` if you wanted.  See also https://github.com/gatsbyjs/gatsby/tree/master/examples

**Reply by yidongw** [reply]
Thank you, but now I have some problems. 

The following error will show up when I delete my `.cache` folder, and then do `gatsby develop` or `gatsby build` 
```
[gatsby-plugin-ts-config] An error occurred while reading your gatsby-node!
[BABEL] /Users/alanwang/myProject/gatsby/gatsby-node.ts: `babel-preset-gatsby` has been loaded, which consumes
config generated by the Gatsby CLI. Set `NODE_ENV=test` to bypass, or run `gatsby build` first. (While processing:
"/Users/alanwang/myProject/node_modules/babel-preset-gatsby/index.js")
```

my `.babelrc` file looks like this
```
{
  "presets": ["babel-preset-gatsby"],
  "plugins": [
    "babel-plugin-styled-components",
    ["@babel/plugin-proposal-decorators", { "legacy": true }],
    ["@babel/plugin-proposal-class-properties", { "loose": true }]
  ]
}
```

After I recover my `.cache` folder from `Trash` then it started working again for `gatsby develop`

-----
> const getDataFp = (cache) => path.join(cache.directory, "data.json");

`cache` doesn't have directory, so I'm using `path.resolve(__dirname, "../.cache/data.json")` instead

**Reply by Js-Brecht** [reply]
> The following error will show up when I delete my `.cache` folder, and then do `gatsby develop` or `gatsby build`

That's because you've defined your own `.babelrc`.  It generally works better if you use the [onCreateBabelConfig](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#onCreateBabelConfig) api.  More specifically (and in your case), the [setBabelPlugin](https://www.gatsbyjs.com/docs/reference/config-files/actions/#setBabelPlugin) action inside of that api

> `cache` doesn't have directory, so I'm using `path.resolve(__dirname, "../.cache/data.json")` instead

Where are you using `cache` from?  It should definitely have `.directory`.  I know it always has when I wanted to use it, including now (Gatsby v3.1.3)

I do stuff kind of like this to inspect the api sometimes.  I don't actually use the `debugger` symbol, but it conveys the meaning pretty well.  If you use a debug console and run `gatsby build` from it, it should break inside of `sourceNodes` and you can view parameters the api is provided.
```js
exports.sourceNodes = async ({ cache }) => {
    console.log(cache.directory);
    debugger;
};
```

**Reply by Js-Brecht** [reply]
Btw, there's [gatsby-plugin-styled-components](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-styled-components) if you wanted to use `styled-components`.  It does a little extra work for you, too

**Reply by yidongw** [reply]
cache's type is defined in https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/index.d.ts but it doesn't have `directory`. I looked it up and did see `directory` inside the cache. Maybe the type is wrong?
```
export interface GatsbyCache {
  /**
   * Retrieve cached value
   * @param key Cache key
   * @returns Promise resolving to cached value
   * @example
   * const value = await cache.get(`unique-key`)
   */
  get(key: string): Promise<any>

  /**
   * Cache value
   * @param key Cache key
   * @param value Value to be cached
   * @returns Promise resolving to cached value
   * @example
   * await cache.set(`unique-key`, value)
   */
  set(key: string, value: any): Promise<any>
}
```

Babel error is now fixed. Thank you!

**Reply by Js-Brecht** [reply]
https://github.com/gatsbyjs/gatsby/blob/060f7321ee84bf540e8fb582273a945d8a27cc22/packages/gatsby/index.d.ts#L1261-L1282

**Reply by yidongw** [reply]
oh. I need to update my gatsby then. Thanks
